src/main/java/frc/robot/subsystems/SwerveSubsystem.java

```diff
@@ -0,0 +1,191 @@
+        gyro.reset();
+    }
+
+    public double getHeading() {
+        //The math for this remainder is gyro - (360 * Math.round(gyro/360))
+        return Math.IEEEremainder(gyro.getAngle(), 360); //Clamps angle output between -180 and 180 degrees
+    }
+
+    public Rotation2d getRotation2d() {
+        return Rotation2d.fromDegrees(getHeading());
+    }
+
+    public Pose2d getPose() {
+        return odometer.getPoseMeters();
+    }
+
+    public void resetOdometry(Pose2d pose) {
+        odometer.resetPosition(pose, getRotation2d());
+    }
+
+    public void toggleOrientation(){
+        //Toggle control orientation from FOD/ROD
+        controlOrientationIsFOD = !controlOrientationIsFOD;
+        controlOrientationLog.append(controlOrientationIsFOD);
+    }
+
+    public SwerveModuleState[] convertToModuleStates(double xTranslation, double yTranslation, double rotation) {
+        //Takes axis input from joysticks and returns an array of swerve module states
+
+        double x = yTranslation; //Intentional, x in swerve kinematics is y on the joystick
+        double y = xTranslation;
+        double r = rotation;
+
+        //Map to speeds in meters/radians per second
+        x *= DriveConstants.kPhysicalMaxSpeed;
+        y *= DriveConstants.kPhysicalMaxSpeed;
+        r *= DriveConstants.kPhysicalMaxAngularSpeed;
+
+        //Log speeds used to construct chassis speeds
+        translationXOutputLog.append(x);
+        translationYOutputLog.append(y);
+        rotationOutputLog.append(r);
+
+        //Construct Chassis Speeds
+        ChassisSpeeds chassisSpeeds;
+        if(controlOrientationIsFOD){
+            //Field Oriented Drive
+            //TODO: test on real hardware, adding to gyro is supposed to eliminate skew of movement when rotating
+            chassisSpeeds = ChassisSpeeds.fromFieldRelativeSpeeds(x, y, r, this.getRotation2d().plus(new Rotation2d((r * Robot.kDefaultPeriod)/2)));
+        } else {
+            //Robot Oriented Drive
+            chassisSpeeds = new ChassisSpeeds(x, y, r);
+        }
+        SmartDashboard.putString("chassis speeds",chassisSpeeds.toString());
+        //Convert Chassis Speeds to individual module states
+        return DriveConstants.kDriveKinematics.toSwerveModuleStates(chassisSpeeds);  
+    }
+
+    private void normalizeDrive(SwerveModuleState[] desiredStates, ChassisSpeeds speeds){
+        //Find magnitude of translation input, map to a scale of 1 in order to be comparable to rotation
+        double translationalK = Math.hypot(speeds.vxMetersPerSecond, speeds.vyMetersPerSecond) / DriveConstants.kPhysicalMaxSpeed;
+        //Find magnitude of rotation input, map to a scale of 1 in order to be comparable to translation
+        double rotationalK = Math.abs(speeds.omegaRadiansPerSecond) / DriveConstants.kPhysicalMaxAngularSpeed;
+        //Use the larger of the two magnitudes for scaling
+        double k = Math.max(translationalK, rotationalK);
```

Comment:
Should the rotationalK be a factor here? Both values have the same normalized range, but are still not the same units. i.e. rotationalK scaled by kPhysicalMaxSpeed isn't meaningful, unless there is some physics I'm missing.

Reply:
rotationalK is scaled by kPhysicalMaxAngularSpeed, which is in radians

However, I don't know how to handle this method, because it's not working in simulation, commanding modules with speeds over the maximum when the drive moves diagonally. (I tried reverting to WPI desaturation as well, but that did the same thing)

Reply:
It gets scaled using kPhysicalMaxSpeed on line 150